### DISCUSSION TITLE
Null bytes in strings

### DISCUSSION THREAD

**Discussion by apollo13** [psycopg/psycopg#385] [General]
While working on the database backend for Django I ran over https://github.com/psycopg/psycopg/blob/40346c8c4518b755c63ece12a3acc1de86521b8e/psycopg/psycopg/types/string.py#L39-L40 

I am wondering if this is actually necessary to have, wouldn't the server throw an error anyways if you were to submit such a query? Some parts of Django compose queries locally (mainly for logging) and it would be nice if we were able to compose the query even if it then fails on the server.

**Comment by dvarrazzo** [comment]
Postgres treats the null in string differently in text and binary format. IIRC it raises an exception in binary format and silently truncate data in text. Raising the exception there makes the two modes behave consistently.

Postgres doesn't allow nulls in strings and we don't want to register a default dumper which allows for invalid data to be passed to the server and silently corrupted. This might hide application errors.

However, psycopg can be customized. You can subclass the `StrDumperUnknown` and avoid to raise an exception in `dump()`. You can use it only in specific places or globally in your application. If you consider it a good idea, such customisation can be made in the psycopg 3 Django backend.

Info is [in the docs](https://www.psycopg.org/psycopg3/docs/advanced/adapt.html), my Django backend implementation already performs quite an extensive customisation of the adaptation map, and you could definitely add to it.

**Comment by apollo13** [comment]
Oh, I didn't realize it silently truncates. That makes the error certainty the better option. Thank you!
